src/main/java/minijava/ir/assembler/AssemblerGenerator.java

```diff
@@ -0,0 +1,584 @@
+package minijava.ir.assembler;
+
+import static minijava.ir.utils.FirmUtils.getMethodLdName;
+import static minijava.ir.utils.FirmUtils.isPhiProneToLostCopies;
+
+import com.sun.jna.Platform;
+import firm.BackEdges;
+import firm.Graph;
+import firm.Mode;
+import firm.Program;
+import firm.nodes.*;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+import minijava.ir.DefaultNodeVisitor;
+import minijava.ir.NameMangler;
+import minijava.ir.assembler.block.AssemblerFile;
+import minijava.ir.assembler.block.CodeBlock;
+import minijava.ir.assembler.block.CodeSegment;
+import minijava.ir.assembler.block.Segment;
+import minijava.ir.assembler.instructions.*;
+import minijava.ir.assembler.instructions.Add;
+import minijava.ir.assembler.instructions.And;
+import minijava.ir.assembler.instructions.Call;
+import minijava.ir.assembler.instructions.Cmp;
+import minijava.ir.assembler.instructions.Div;
+import minijava.ir.assembler.instructions.Jmp;
+import minijava.ir.assembler.instructions.Mul;
+import minijava.ir.assembler.instructions.Sub;
+import minijava.ir.assembler.location.Location;
+import minijava.ir.assembler.location.Register;
+import minijava.ir.assembler.location.StackLocation;
+import minijava.ir.utils.MethodInformation;
+
+/**
+ * Generates GNU assembler for a graph
+ *
+ * <p>
+ *
+ * <p>One of the goals of this implementation is to produce well documented assembly code.
+ *
+ * <p>
+ *
+ * <p>Important: it currently ignores any {@link firm.nodes.Conv} nodes
+ */
+public class AssemblerGenerator implements DefaultNodeVisitor {
+
+  private final Graph graph;
+  private final MethodInformation info;
+  private final NodeAllocator allocator;
+  private final CodeSegment segment;
+  private Map<Integer, CodeBlock> blocksToCodeBlocks;
+
+  public AssemblerGenerator(Graph graph, NodeAllocator allocator) {
+    this.graph = graph;
+    this.info = new MethodInformation(graph);
+    this.allocator = allocator;
+    allocator.process(graph);
+    blocksToCodeBlocks = new HashMap<>();
+    segment = new CodeSegment(new ArrayList<>(), new ArrayList<>());
+    segment.addComment(String.format("Code segment for method %s", info.name));
+    blocksToCodeBlocks = new HashMap<>();
+  }
+
+  public Segment generate() {
```

Comment:
Generate... what exactly? Judging from the implementation this generates Code for the passed graph.

Hmm... Why not make the graph a parameter of the method? That would be a lot clearer. I really don't like (unnecessary?) instance state like that.